### ISSUE TITLE
Proposal: Endianness Enum and Extensions for BinaryReader an BinaryWriter

### ISSUE THREAD

**Issue by deinok** [dotnet/corefx#31265]
# Proposal
Add an enum that handles the endiannes
```csharp
	/// <summary>
	/// Endianness refers to the sequential order in which bytes are arranged into larger numerical values when stored in memory or when transmitted over digital links.
	/// </summary>
	public enum Endianness {

		/// <summary>
		/// The least significant byte (LSB) value, is at the lowest address. 
		/// </summary>
		LittleEndian,

		/// <summary>
		/// The most significant byte (MSB) value, is at the lowest address.
		/// </summary>
		BigEndian

	}
```

Also add extensions to BinaryReader like:
```csharp
public static ushort ReadUInt16(this BinaryReader binaryReader, Endianness endianness);
public static short ReadInt16(this BinaryReader binaryReader, Endianness endianness);
etc...
```

And for BinaryWriter:
```csharp
public static void Write(this BinaryWriter binaryWriter, ushort value, Endianness endianness);
public static void Write(this BinaryWriter binaryWriter, short value, Endianness endianness);
etc...
```

# Rationale
BinaryReader and BinaryWriter works really well for accessing Streams but they seems to lack the support for byte-order. And the usage of an enum for representing the Endianness seems more clear to me than using the BitConverter.IsLittleEndian bool.

**Comment by khellang** [comment]
Sounds like it would overlap quite a bit with [`BinaryPrimitives`](https://github.com/dotnet/corefx/tree/master/src/System.Memory/src/System/Buffers/Binary)?

**Comment by deinok** [comment]
@khellang Yes, I have implemented this by myself and i have done it with help of BinaryPrimitives.ReverseEndianness();
But as I feel it, seems more clear to use BinaryRW when accessing streams than the way it is done via BinaryPrimitives

**Comment by vcsjones** [comment]
I've not worked with many data streams that need to read different endianness data in the same stream.

`BinaryReader` seems to always read data in little endian format. Some other considerations.

1. What about "I want to use the endianness of whatever the architecture is"?

1. Should this be a class-level thing instead of per-method? It might be tedious and error prone to keep having to pass around the right endianness in different methods.

1. Does it make sense to implement an entirely different type? `BigEndianStreamReader` or something along those lines since the current one appears to always use little?

**Comment by deinok** [comment]
Hi @vcsjones I have never worked on something that needs to change the Endianness in the same stream, but it could be a possible (I cant imagine a case like this).

I'm not completely sure, but I think that BinaryReader takes by default the Endianness of the architecture by default (Most arch are little endian).

1.- As I said, I think its the default (What is actualy implemented). But I agree that would be great to have a static property in some class that represents the system information that handles the Endianess. Just like BitConverter.IsLittleEndian but returning the actual Endianness enum value.

2.- I would prefere the API I have described, but I don't think its a bad idea what you said. Something like  set the default endianness of the BinaryRW on the constructor. I think that we can have the two things at the same time without cost.

3.- As I said, I think that it gets the default of the architecture, so your case can be handled in the example of 2, putting the default endiannes on the constructor or something like that.

**Comment by Halofreak1990** [comment]
> I'm not completely sure, but I think that BinaryReader takes by default the Endianness of the architecture by default (Most arch are little endian).

As far as I'm aware, the .NET CLR itself, is little-endian, and the BinaryReader/Writer implementations are built with this in mind, swapping bytes as necessary when running on a big-endian machine. This simplifies running code on the CLR, as endianness is completely taken out of the equation.

**Comment by deinok** [comment]
@Halofreak1990 If somebody from CoreCLR could provide us a bit more of help

**Comment by JeremyKuhne** [comment]
I'm not opposed to having additional members, but I would want to see more feedback to justify the value here. It is relatively trivial to derive from `BinaryReader` and create an `EndianReader` and your code would be portable.

**Comment by deinok** [comment]
@JeremyKuhne For my experience, most of the times I use the `BinaryReader` and `BinaryWriter` I need to enforce an endianess. In fact, I have never used the `BinaryReader` and `BinaryWriter` without forcing a specific endiannes.

**Comment by Wraith2** [comment]
> Hi @vcsjones I have never worked on something that needs to change the Endianness in the same stream, but it could be a possible (I cant imagine a case like this).

Just a data point on this,  parts of the TDS protocol used in SqlClient require specific and opposite endian-ness

**Comment by Ryder25** [comment]
For example, if you have a game data file that needs to be written using different endianness depending on the platform.

If reading/writing the file for Windows, then you may use LittleEndian, and instantiate the BinaryReader/Writer as so.  Whereas for the PlayStation 3, the reader/writer would be instantiated with BigEndian.

**Comment by Halofreak1990** [comment]
@Ryder25 that is assuming you'd output your assets in a different endianness for each specific platform, which is something I'd personally never do. Why complicate development, if the framework automagically fixes things for you? .NET, as it currently works, allows one to use little-endian assets, and in code never care about that, unless you need to fiddle with specific bits. For example, a little endian long value in a file will produce the expected value in code, regardless of the endianness of the platform. I'd be inclined to say that, if you really need control of the endiannes on a case by case basis, you really need to re-evaluate what you're doing, because something's out of whack.

**Comment by Ryder25** [comment]
I'm not saying I would do it, but I've seen it done before with several games. I can see a case where the game itself is written in C++, and there could be tooling done in C#. At that point, you can't only focus on what's happening in the .Net bcl.

**Comment by carlossanlop** [comment]
Triage: There are enough scenarios that we see the value in doing this.
